js/test/test.js

```diff
@@ -296,165 +373,212 @@ async function testExchange (exchange) {
+    // this might not be ideal to handle all markettypes (inverse, futures, options, etc...) but we can't aim at this stage to cover everything, and let's just focus on spot and swap for this PR
+    const isSymbolProvided = providedSymbol;
+    if (providedSymbol) {
+        if (providedSymbol.indexOf(':') < 0) {
+            spotSymbol = providedSymbol;
+        } else {
+            swapSymbol = providedSymbol;
+        }
+    } else {
+        spotSymbol = getValidSymbol (exchange);
+        swapSymbol = getValidSymbol (exchange, false);
+    }
+    if (spotSymbol !== undefined) {
+        console.log ('SPOT SYMBOL:', spotSymbol);
+    }
+    if (swapSymbol !== undefined) {
+        console.log ('SWAP SYMBOL:', swapSymbol);
     }
 
-    console.log ('SYMBOL:', symbol);
-    if ((symbol.indexOf ('.d') < 0) && !privateOnly) {
-        await testSymbol (exchange, symbol);
+    if (!privateOnly) {
+        if (exchange.has['spot']) {
+            exchange.options['type'] = 'spot';
+            await runPublicTests (exchange, spotSymbol);
+        }
+        if (exchange.has['swap']) {
+            exchange.options['type'] = 'swap';
+            await runPublicTests (exchange, swapSymbol);
+        }
     }
 
     if (privateTest || privateOnly) {
-        if (!exchange.privateKey && (!exchange.apiKey || (exchange.apiKey.length < 1))) {
-            console.log ('[Skipped]', 'Keys not found, skipping private tests');
-            return true;
+        if (exchange.has['spot']) {
+            exchange.options['type'] = 'spot';
+            await runPrivateTests (exchange, spotSymbol);
+        }
+        if (exchange.has['swap']) {
+            exchange.options['type'] = 'swap';
+            await runPrivateTests (exchange, swapSymbol);
         }
-        await runPrivateTests (exchange, symbol);
     }
 }
 
 //-----------------------------------------------------------------------------
 
 async function runPrivateTests(exchange, symbol) {
-    exchange.checkRequiredCredentials ();
-
-    await test ('signIn', exchange);
-
-    const code = getExchangeCode (exchange);
-
-    const balance = await test ('fetchBalance', exchange);
-
-    await test ('fetchAccounts', exchange);
-    await test ('fetchTransactionFees', exchange);
-    await test ('fetchTradingFees', exchange);
-    await test ('fetchStatus', exchange);
-
-    await test ('fetchOrders', exchange, symbol);
-    await test ('fetchOpenOrders', exchange, symbol);
-    await test ('fetchClosedOrders', exchange, symbol);
-    await test ('fetchMyTrades', exchange, symbol);
-    await test ('fetchLeverageTiers', exchange, symbol);
-    await test ('fetchOpenInterestHistory', exchange, symbol);
-
-    await test ('fetchPositions', exchange, symbol);
-
-    if ('fetchLedger' in tests) {
-        await test ('fetchLedger', exchange, code);
+    if (!exchange.privateKey && (!exchange.apiKey || (exchange.apiKey.length < 1)) || exchange.checkRequiredCredentials (false)) {
```

Comment:
@ttodua usually, adding complexity to the conditions is not a good idea, adding `checkRequiredCredentials` like that, breaks this check because it returns true if the keys are detected. It should be enough to use `checkRequiredCredentials` tho` (going to change it)

Reply:
ouch, i had a type. missed `!` in front of `checkRequiredCredentials()`.